opm/autodiff/ISTLSolverEbos.hpp

```diff
@@ -381,11 +622,255 @@ namespace Opm
+	  auto& bx = x[i];
+	  for(std::size_t jj=0; jj < bx.size(); jj++){
+	    double var_scale = simulator_.model().primaryVarWeight(i,jj);
+            bx[jj] /= var_scale;
+	  }	  
+	}
+      }
+      
+      Vector getQuasiImpesWeights(){
+	Matrix& A = *matrix_;
+	Vector weights(rhs_->size());
+	BlockVector rhs(0.0);
+	rhs[pressureIndex] = 1;
+	const auto endi = A.end();
+	int index = 0;
+	for (auto i=A.begin(); i!=endi; ++i){
+	  const auto endj = (*i).end();                                        
+	  MatrixBlockType diag_block(0.0);
+	  for (auto j=(*i).begin(); j!=endj; ++j){
+	    if(i.index() == j.index()){
+	      diag_block = (*j);
+	      break;
+	    }                        
+	  }                    
+	  BlockVector bweights;
+	  auto diag_block_transpose = diag_block.transpose();
+	  diag_block_transpose.solve(bweights, rhs);
+	  double abs_max =
+	    *std::max_element(bweights.begin(), bweights.end(), [](double a, double b){ return std::abs(a) < std::abs(b); } );
+	  bweights /= std::abs(abs_max);
+	  weights[i.index()] = bweights;
+	}
+	return weights;
+      }
+      
+      Vector getSimpleWeights(const BlockVector& rhs){
+	Vector weights(rhs_->size(),0);
+	for(auto& bw: weights){
+	  bw = rhs;
+	}
+	return weights;
+      }
+
+      void scaleMatrixAndRhs(const Vector& weights){
+	static_assert(pressureIndex == 0, "Current support that pressure equation should be first");
+	//using Matrix = typename Operator::matrix_type;
+	using Block = typename Matrix::block_type;
+	//Vector& rhs = *rhs_;
+	//Matrix& A = *matrix_;
+	//int index = 0;
+	//for ( auto& row : *matrix_ ){
+	const auto endi = matrix_->end();
+	for (auto i=matrix_->begin(); i!=endi; ++i){
+	  
+	  //const auto& bweights = weights[row.index()];
+	  const BlockVector& bweights = weights[i.index()];
+	  BlockVector& brhs = (*rhs_)[i.index()];
+	  //++index;
+	  //for ( auto& block : row ){
+	  const auto endj = (*i).end();                                        
+	  for (auto j=(*i).begin(); j!=endj; ++j){    
+	    // assume it is something on all rows
+	    // the blew logic depend on pressureIndex=0
+	    Block& block = (*j);
+	    for ( std::size_t ii = 0; ii < block.rows; ii++ ){
+	      if ( ii == 0 ){
+		for(std::size_t jj=0; jj < block.cols; jj++){
+		  block[0][jj] *= bweights[ii];//*block[ii][jj];
+		}
+	      } else {
+		for(std::size_t jj=0; jj < block.cols; jj++){
+		  block[0][jj] += bweights[ii]*block[ii][jj];
+		}
+	      }
+	    }
+	    
+	  }
+	  for(std::size_t ii=0; ii < brhs.size(); ii++){
+	    if ( ii == 0 ){
+	      brhs[0] *= bweights[ii];//*brhs[ii];
+	    }else{
+	      brhs[0] += bweights[ii]*brhs[ii];
+	    }
+	  }	      
+	}
+      }
+      
+      static void multBlocksInMatrix(Matrix& ebosJac,const MatrixBlockType& trans,bool left=true){
+	const int n = ebosJac.N();
+	//const int np = FluidSystem::numPhases;
+	for (int row_index = 0; row_index < n; ++row_index) {
+	  auto& row = ebosJac[row_index];
+	  auto* dataptr = row.getptr();
```

Comment:
Is this a more C-like manner of doing
```
for( auto&& block: row){
  if(left)
    block = block.leftmultiply(trans);
  else
    block = block.rightmultiply(trans);
}
```
Or am I missing the reason why we need to convert to a pointer?